pkg/reconciler/eventtype/eventtype.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eventtype
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Only allow alphanumeric, '-' or '.'.
+var validChars = regexp.MustCompile(`[^-\.a-z0-9]+`)
+
+// Reconciler reconciles EventTypes of different sources.
+type Reconciler struct {
```

Comment:
Is this something that can be pulled in from eventing?

Reply:
I may be able to put it there, as we might need it for the core sources as well. 
Do you want me to do that? 
The downside with that is that we will have a dependency to eventing of other things than just API types. 
@grantr and I were thinking of pushing for an api repo, and only depend on that in the future. I guess that will take some time to get approved, so we might be able to move this there.

Reply:
Actually, I think it would be better to have this stuff here (and possibly do @grantr changes). 
The only source from eventing that might need to create EventTypes is CronJob, and it can do it as we typically do when creating resources from controllers. If we later on see a need for this in eventing, I can move it there.  
If there is a strong objection, we can further discuss. I'll mark this as resolve for now